gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/converter/EnvelopeSchemaConverter.java

```diff
@@ -17,136 +17,141 @@
+          throws Exception {
+        return (Schema) registry.getSchemaByKey(key);
       }
     });
     return this;
   }
 
   /**
-   * Do nothing, actual schema must be obtained from records.
-   */
-  @Override
-  public String convertSchema(Schema inputSchema, WorkUnitState workUnit) throws SchemaConversionException {
-    return EnvelopeSchemaConverter.class.getName();
-  }
-
-  /**
-   * Get actual schema from registry and deserialize payload using it.
+   * Get the payload schema
+   *
+   * @param inputRecord the input record which has the payload
+   * @return the current schema of the payload
    */
-  @Override
-  public Iterable<GenericRecord> convertRecord(String outputSchema, GenericRecord inputRecord, WorkUnitState workUnit)
-      throws DataConversionException {
-    try {
-      String schemaIdField = workUnit.contains(PAYLOAD_SCHEMA_ID_FIELD) ?
-          workUnit.getProp(PAYLOAD_SCHEMA_ID_FIELD) : DEFAULT_PAYLOAD_SCHEMA_ID_FIELD;
-      String payloadField = workUnit.contains(PAYLOAD_FIELD) ?
-          workUnit.getProp(PAYLOAD_FIELD) : DEFAULT_PAYLOAD_FIELD;
-      String schemaKey = String.valueOf(inputRecord.get(schemaIdField));
-      Schema payloadSchema = (Schema) this.registry.getSchemaByKey(schemaKey);
-      byte[] payload = getPayload(inputRecord, payloadField);
-      GenericRecord outputRecord = deserializePayload(payload, payloadSchema);
-      if (this.fieldRemover.isPresent()) {
-        payloadSchema = this.fieldRemover.get().removeFields(payloadSchema);
-      }
-      return new SingleRecordIterable<>(AvroUtils.convertRecordSchema(outputRecord, payloadSchema));
-    } catch (IOException | SchemaRegistryException | ExecutionException e) {
-      throw new DataConversionException(e);
+  protected Schema getPayloadSchema(GenericRecord inputRecord)
+      throws Exception {
+    Optional<Object> schemaIdValue = AvroUtils.getFieldValue(inputRecord, payloadSchemaIdField);
+    if (!schemaIdValue.isPresent()) {
+      throw new Exception("Schema id with key " + payloadSchemaIdField + " not found in the record");
     }
+    String schemaKey = String.valueOf(schemaIdValue.get());
+    return schemaCache.get(schemaKey);
   }
 
   /**
-   * Get payload field from GenericRecord and convert to byte array
+   * Get payload field and convert to byte array
+   *
+   * @param inputRecord the input record which has the payload
+   * @return the byte array of the payload in the input record
    */
-  public byte[] getPayload(GenericRecord inputRecord, String payloadFieldName) {
-    ByteBuffer bb = (ByteBuffer) inputRecord.get(payloadFieldName);
-    byte[] payloadBytes;
+  protected byte[] getPayloadBytes(GenericRecord inputRecord) {
+    ByteBuffer bb = (ByteBuffer) inputRecord.get(payloadField);
     if (bb.hasArray()) {
-      payloadBytes = bb.array();
+      return bb.array();
     } else {
-      payloadBytes = new byte[bb.remaining()];
+      byte[] payloadBytes = new byte[bb.remaining()];
```

Comment:
Maybe the childclass should overwrite this method instead of putting if/else in the abstract class?

Reply:
The childclass can always overwrite this method as needed. Here is just a default implementation of the method.